File: src/main/scala/com/yogpc/qp/machines/mini_quarry/MiniQuarryTile.scala
package com.yogpc.qp.machines.mini_quarry

import java.util

import cats.implicits._
import com.mojang.datafixers.Dynamic
import com.yogpc.qp._
import com.yogpc.qp.compat.InvUtils
import com.yogpc.qp.machines.base._
import com.yogpc.qp.machines.modules.{IModuleItem, ItemFuelModule}
import com.yogpc.qp.machines.quarry.QuarryFakePlayer
import com.yogpc.qp.machines.{PowerManager, TranslationKeys}
import com.yogpc.qp.packet.{PacketHandler, TileMessage}
import com.yogpc.qp.utils.Holder
import net.minecraft.block.Block
import net.minecraft.entity.player.{PlayerEntity, PlayerInventory}
import net.minecraft.inventory.container.{Container, INamedContainerProvider}
import net.minecraft.inventory.{IInventory, InventoryHelper, ItemStackHelper}
import net.minecraft.item.ItemStack
import net.minecraft.nbt.{CompoundNBT, NBTDynamicOps}
import net.minecraft.state.properties.BlockStateProperties
import net.minecraft.util.math.{AxisAlignedBB, BlockPos}
import net.minecraft.util.text.{ITextComponent, StringTextComponent}
import net.minecraft.util.{Direction, Hand, NonNullList, ResourceLocation}
import net.minecraft.world.server.ServerWorld
import net.minecraftforge.api.distmarker.{Dist, OnlyIn}
import net.minecraftforge.common.ForgeHooks
import net.minecraftforge.common.util.Constants.NBT

import scala.collection.AbstractIterator
import scala.jdk.CollectionConverters._

class MiniQuarryTile extends APowerTile(Holder.miniQuarryType)
  with EnchantmentHolder.EnchantmentProvider
  with IEnchantableTile
  with IRemotePowerOn
  with IChunkLoadTile
  with INamedContainerProvider
  with IDebugSender {
  private final var enchantments = EnchantmentHolder.noEnch
  private final var area = Area.zeroArea
  private final var targets: List[BlockPos] = List.empty
  private final val tools = NonNullList.withSize(5, ItemStack.EMPTY)
  private final var blackList: Set[QuarryBlackList.Entry] = Set(QuarryBlackList.Air)
  private final var preDirection: Direction = Direction.UP
  final var rs = false
  final var renderBox = false
  private[this] final val dropItem = (item: ItemStack) => {
    val rest = InvUtils.injectToNearTile(world, pos, item)
    InventoryHelper.spawnItemStack(world, pos.getX + 0.5, pos.getY + 1, pos.getZ + 0.5, rest)
  }

  override protected def workInTick(): Unit = {
    if (world.getGameTime % MiniQuarryTile.interval(enchantments.efficiency) == 0 &&
      PowerManager.useEnergy(this, MiniQuarryTile.e(enchantments.unbreaking), EnergyUsage.MINI_QUARRY)) {
      // Work
      @scala.annotation.tailrec
      def work(poses: List[BlockPos]): List[BlockPos] = {
        poses match {
          case head :: tail =>
            val world = getDiggingWorld
            val state = world.getBlockState(head)
            if (blackList.exists(_.test(state, world, head)) ||
              state.getBlockHardness(world, head) < 0) {
              // Unbreakable
              work(tail)
            } else {
              // Check if block harvest-able.
              val fakePlayer = QuarryFakePlayer.get(world, head)
              val canHarvest = tools.asScala.filterNot(_.getItem.isInstanceOf[IModuleItem]).exists { tool =>
                fakePlayer.setHeldItem(Hand.MAIN_HAND, tool)
                ForgeHooks.canHarvestBlock(state, fakePlayer, world, head) || ForgeHooks.isToolEffective(world, head, tool)
              }
              // Use effective tool
              tools.asScala.filterNot(_.getItem.isInstanceOf[IModuleItem]).find(tool => ForgeHooks.isToolEffective(world, head, tool))
                .foreach(fakePlayer.setHeldItem(Hand.MAIN_HAND, _))
              if (canHarvest) {
                // Remove block
                val drops = Block.getDrops(state, world, head, world.getTileEntity(head), fakePlayer, fakePlayer.getHeldItemMainhand)
                drops.asScala.foreach(dropItem)
                val count = if (ForgeHooks.isToolEffective(world, head, fakePlayer.getHeldItemMainhand)) 1 else 4
                for (_ <- 0 until count)
                  fakePlayer.getHeldItemMainhand.onBlockDestroyed(world, state, head, fakePlayer)
                world.removeBlock(head, false)
                tail
              } else {
                // Skip this block
                tail
              }
            }
          case Nil => Nil
        }
      }

      targets = work(targets)
      if (targets.isEmpty)
        updateWorkingState()
    }
  }

  override protected def getEnergyInTick(): Unit = {
    tools.asScala.map(i => (i.getItem, i)).collectFirst { case (f: ItemFuelModule, s) => f.getFuelModule(s) }.foreach {
      f => f.invoke(IModule.Tick(this))
    }
  }

  def getDiggingWorld: ServerWorld = {
    if (!super.getWorld.isRemote) {
      this.area.getWorld(super.getWorld.asInstanceOf[ServerWorld])
    } else {
      throw new IllegalStateException("Tried to get server world in client.")
    }
  }

  def gotRSPulse(): Unit = {
    if (isWorking) {
      startWaiting()
    } else {
      startWorking()
    }
  }

  def getInv: IInventory = Inv

  @OnlyIn(Dist.CLIENT)
  def renderAreaBox: Boolean = getBlockState.get(QPBlock.WORKING) && renderBox

  override protected def isWorking: Boolean = targets.nonEmpty

  override def getEnchantmentHolder: EnchantmentHolder = enchantments

  override def G_ReInit(): Unit = {
    PowerManager.configureQuarryWork(this, 0, 0, 0)
  }

  override def getEnchantments: util.Map[ResourceLocation, Integer] =
    EnchantmentHolder.getEnchantmentMap(enchantments).collect(enchantCollector).asJava

  override def setEnchantment(id: ResourceLocation, level: Short): Unit = enchantments =
    EnchantmentHolder.updateEnchantment(enchantments, id, level)

  override def setArea(area: Area): Unit = this.area = area

  override def startWorking(): Unit = {
    val facing = getBlockState.get(BlockStateProperties.FACING)
    val maybeMarkers = Area.getMarkersOnDirection(List(facing.getOpposite, facing.rotateY(), facing.rotateYCCW()), world, pos, ignoreHasLink = true)
    val areas = maybeMarkers.map(m => Area.posToArea(m.min(), m.max(), world.getDimension.getType) -> m)
      .collectFirst(t => t)
    areas match {
      case Some((newArea, m)) =>
        area = newArea
        m.removeFromWorldWithItem().asScala.foreach(dropItem)
      case _ =>
    }
    updateTargets(facing.getOpposite)
    updateWorkingState()
    if (!world.isRemote) PacketHandler.sendToClient(TileMessage.create(this), world)
  }

  private def updateTargets(d: Direction): Unit = {
    if (area == Area.zeroArea) {
      targets = List.empty
    } else {
      preDirection = d
      val poses = for {
        y <- area.yMax to(area.yMin, -1)
        (x, z) <- MiniQuarryTile.makeTargetsXZ(area, d)
      } yield new BlockPos(x, y, z)
      targets = poses.toList
    }
  }

  override def startWaiting(): Unit = {
    targets = List.empty
    updateWorkingState()
  }

  override def getArea: Area = area

  override def createMenu(id: Int, i: PlayerInventory, player: PlayerEntity): Container = new MiniQuarryContainer(id, player, pos)

  override def getDebugName: String = TranslationKeys.mini_quarry

  override def getDebugMessages: util.List[_ <: ITextComponent] = Seq(
    s"Area: $area}",
    s"TargetSize: ${targets.size}",
  ).map(new StringTextComponent(_)).asJava

  override def getDisplayName: ITextComponent = super.getDisplayName

  override def remove(): Unit = {
    super[IChunkLoadTile].releaseTicket()
    super.remove()
  }

  override def read(nbt: CompoundNBT): Unit = {
    super.read(nbt)
    this.area = Area.areaLoad(nbt.getCompound("area"))
    this.enchantments = EnchantmentHolder.enchantmentHolderLoad(nbt, "enchantments")
    this.preDirection = Direction.byName(nbt.getString("preDirection"))
    val working = nbt.contains("head")
    if (working) {
      updateTargets(preDirection)
      val head = BlockPos.fromLong(nbt.getLong("head"))
      this.targets = targets.dropWhile(_ == head)
    }
    ItemStackHelper.loadAllItems(nbt.getCompound("tools"), tools)
    this.blackList = nbt.getList("blackList", NBT.TAG_COMPOUND).asScala
      .map(n => QuarryBlackList.readEntry(new Dynamic(NBTDynamicOps.INSTANCE, n))).toSet
    this.rs = nbt.getBoolean("rs")
    this.renderBox = nbt.getBoolean("renderBox")
  }

  override def write(nbt: CompoundNBT): CompoundNBT = {
    nbt.put("area", area.toNBT)
    nbt.put("enchantments", enchantments.toNBT)
    nbt.putString("preDirection", preDirection.getName)
    targets.headOption.foreach(p => nbt.putLong("head", p.toLong))
    nbt.put("tools", ItemStackHelper.saveAllItems(new CompoundNBT(), tools))
    nbt.put("blackList", NBTDynamicOps.INSTANCE.createList(blackList.asJava.stream().map(QuarryBlackList.writeEntry(_, NBTDynamicOps.INSTANCE))))
    nbt.putBoolean("rs", rs)
    nbt.putBoolean("renderBox", renderBox)
    super.write(nbt)
  }

  override def hasFastRenderer: Boolean = true

  override def getRenderBoundingBox: AxisAlignedBB = {
    if (area != Area.zeroArea) Area.areaBox(area)
    else super.getRenderBoundingBox
  }

  override def getMaxRenderDistanceSquared: Double = {
    if (area != Area.zeroArea) Area.areaLengthSq(area)
    else super.getMaxRenderDistanceSquared
  }

  override protected def enabledByRS = true

  private object Inv extends IInventory {
    override def getSizeInventory: Int = tools.size()

    override def isEmpty: Boolean = tools.stream().allMatch(_.isEmpty)

    override def getStackInSlot(index: Int): ItemStack = tools.get(index)

    override def decrStackSize(index: Int, count: Int): ItemStack = ItemStackHelper.getAndSplit(tools, index, count)

    override def removeStackFromSlot(index: Int): ItemStack = ItemStackHelper.getAndRemove(tools, index)

    override def setInventorySlotContents(index: Int, stack: ItemStack): Unit = tools.set(index, stack)

    override def markDirty(): Unit = MiniQuarryTile.this.markDirty()

    override def isUsableByPlayer(player: PlayerEntity): Boolean = true

    override def clear(): Unit = tools.clear()
  }

}

object MiniQuarryTile {
  final val SYMBOL = Symbol("MiniQuarry")

  def e(unbreaking: Int): Long = APowerTile.FEtoMicroJ * 20 / (unbreaking + 1)

  def interval(efficiency: Int): Int = {
    if (efficiency < 0) return 100
    efficiency match {
      case 0 => 40
      case 1 => 30
      case 2 => 20
      case 3 => 10
      case 4 => 5
      case 5 => 2
      case _ => 1
    }
  }

  def makeTargetsXZ(area: Area, direction: Direction): Iterator[(Int, Int)] = {
    val (start, end) = direction match {
      case Direction.NORTH | Direction.DOWN | Direction.UP => (area.xMax, area.zMin) -> (area.xMin, area.zMax)
      case Direction.SOUTH => (area.xMin, area.zMax) -> (area.xMax, area.zMin)
      case Direction.WEST => (area.xMin, area.zMin) -> (area.xMax, area.zMax)
      case Direction.EAST => (area.xMax, area.zMax) -> (area.xMin, area.zMin)
    }
    val vec@(vx, vz) = end |-| start
    val (vec1, vec2) = direction.getAxis match {
      case Direction.Axis.X => vec.bimap(_.sign, _ => 0) -> vec.bimap(_ => 0, _.sign)
      case Direction.Axis.Z => vec.bimap(_ => 0, _.sign) -> vec.bimap(_.sign, _ => 0)
      case _ => (0, 1) -> (-1, 0)
    }
    val dot1 = vx * vec1._1 + vz * vec1._2
    val dot2 = vx * vec2._1 + vz * vec2._2
    new AbstractIterator[(Int, Int)] {
      var count1 = 0
      var count2 = 0
      var end = false

      override def hasNext: Boolean = !end

      override def next(): (Int, Int) = {
        val result = start |+| (vec1 combineN count1) |+| (vec2 combineN count2)
        if (count1 + 1 > dot1) {
          count1 = 0
          count2 += 1
          if (count2 == dot2 + 1) end = true
        } else {
          count1 += 1
        }
        result
      }
    }
  }
}

File: src/main/scala/com/yogpc/qp/machines/mini_quarry/MiniQuarryTile.java
package com.yogpc.qp.machines.mini_quarry;

import com.yogpc.qp.Holder;
import com.yogpc.qp.machines.*;
import com.yogpc.qp.utils.MapMulti;
import net.minecraft.ChatFormatting;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.nbt.Tag;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.Container;
import net.minecraft.world.Containers;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraftforge.common.ForgeHooks;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.level.BlockEvent;
import net.minecraftforge.registries.ForgeRegistries;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class MiniQuarryTile extends PowerTile implements CheckerLog,
    EnchantmentLevel.HasEnchantments, MenuProvider, PowerConfig.Provider {
    private List<EnchantmentLevel> enchantments;
    @Nullable
    Area area = null;
    boolean rs;
    final MiniQuarryInventory container = new MiniQuarryInventory();
    @Nullable
    MiniTarget targetIterator;
    Collection<BlockStatePredicate> denyList = defaultBlackList();
    Collection<BlockStatePredicate> allowList = Set.of();

    public MiniQuarryTile(BlockPos pos, BlockState state) {
        super(Holder.MINI_QUARRY_TYPE, pos, state);
        container.addListener(c -> this.setChanged());
    }

    void work() {
        container.getEnergyModule().ifPresent(e -> addEnergy(e.energy() * PowerTile.ONE_FE, false));
        if (!hasEnoughEnergy()) return;
        assert level != null;
        // Interval check
        if (level.getGameTime() % interval(efficiencyLevel()) != 0 || targetIterator == null) return;
        // Energy consumption
        long consumedEnergy = PowerManager.getMiniQuarryEnergy(this);
        if (!useEnergy(consumedEnergy, Reason.MINI_QUARRY, false)) return;
        // Break block
        while (targetIterator.hasNext()) {
            var level = getTargetWorld();
            var pos = targetIterator.next();
            var state = level.getBlockState(pos);
            if (!canBreak(level, pos, state)) {
                TraceQuarryWork.canBreakCheck(this, getBlockPos(), pos, state, "In deny list or unbreakable");
                continue; // The block is in deny list or unbreakable.
            }

            var fakePlayer = QuarryFakePlayer.get(level);
            var event = new BlockEvent.BreakEvent(level, pos, state, fakePlayer);
            if (MinecraftForge.EVENT_BUS.post(event)) {
                TraceQuarryWork.blockRemoveFailed(this, getBlockPos(), pos, state, BreakResult.FAIL_EVENT);
                break; // Denied to break block.
            }

            if (state.getDestroySpeed(level, pos) < 0) {
                // Consume additional energy if quarry tries to remove bedrock.
                var energy = PowerManager.getBreakEnergy(-1, this);
                useEnergy(energy, Reason.BREAK_BLOCK, true);
                consumedEnergy += energy;
            }
            var tools = container.tools();
            var tool = tools.stream().filter(t ->
                t.isCorrectToolForDrops(state)
            ).findFirst().or(() -> tools.stream().filter(t -> {
                fakePlayer.setItemInHand(InteractionHand.MAIN_HAND, t);
                return ForgeHooks.isCorrectToolForDrops(state, fakePlayer);
            }).findFirst());
            final long c = consumedEnergy;
            tool.ifPresent(t -> {
                fakePlayer.setItemInHand(InteractionHand.MAIN_HAND, t);
                var drops = InvUtils.getBlockDrops(state, level, pos, level.getBlockEntity(pos), fakePlayer, t);
                TraceQuarryWork.blockRemoveSucceed(this, getBlockPos(), pos, state, drops, 0, c);
                drops.forEach(this::insertOrDropItem);
                var damage = t.isCorrectToolForDrops(state) ? 1 : 4;
                for (int i = 0; i < damage; i++) {
                    t.mineBlock(level, state, pos, fakePlayer);
                }
                level.removeBlock(pos, false);
                var sound = state.getSoundType();
                level.playSound(null, pos, sound.getBreakSound(), SoundSource.BLOCKS, (sound.getVolume() + 1.0F) / 4F, sound.getPitch() * 0.8F);
            });
            if (tool.isPresent()) break; // Continue if no tools available
        }
        if (!targetIterator.hasNext()) {
            targetIterator = null;
            finishWork();
        }
    }

    boolean canBreak(Level level, BlockPos pos, BlockState state) {
        if (allowList.stream().anyMatch(t -> t.test(state, level, pos))) {
            return true;
        }
        return state.getDestroySpeed(level, pos) >= 0 &&
            denyList.stream().noneMatch(t -> t.test(state, level, pos));
    }

    ServerLevel getTargetWorld() {
        return (ServerLevel) level;
    }

    boolean isWorking() {
        return targetIterator != null;
    }

    void gotRSPulse() {
        if (isWorking()) {
            finishWork();
        } else {
            startWork();
        }
    }

    public void setArea(@Nullable Area area) {
        this.area = area;
        if (area != null)
            this.targetIterator = MiniTarget.of(area, true);
    }

    void startWork() {
        assert level != null;
        var behind = getBlockState().getValue(BlockStateProperties.FACING).getOpposite();
        Area area;
        if (this.area == null) {
            area = Stream.of(behind, behind.getCounterClockWise(), behind.getClockWise())
                .map(getBlockPos()::relative)
                .flatMap(p -> {
                    if (level.getBlockEntity(p) instanceof QuarryMarker marker) return Stream.of(marker);
                    else return Stream.empty();
                })
                .flatMap(m -> m.getArea().stream().peek(a -> m.removeAndGetItems().forEach(this::insertOrDropItem)))
                .findFirst()
                .map(a -> new Area(a.minX() - 1, a.minY(), a.minZ() - 1, a.maxX() + 1, a.maxY(), a.maxZ() + 1, a.direction()))
                .map(a -> a.aboveY(level.getMinBuildHeight() + 1)) // Do not dig the floor of world.
                .orElse(null);
        } else {
            area = this.area;
        }
        setArea(area);
        if (this.area != null) {
            level.setBlock(getBlockPos(), getBlockState().setValue(QPBlock.WORKING, true), Block.UPDATE_ALL);
            TraceQuarryWork.startWork(this, getBlockPos(), getEnergyStored());
        }
    }

    void finishWork() {
        assert level != null;
        this.targetIterator = null;
        level.setBlock(getBlockPos(), getBlockState().setValue(QPBlock.WORKING, false), Block.UPDATE_ALL);
        TraceQuarryWork.finishWork(this, getBlockPos(), this.getEnergyStored());
        logUsage();
    }

    void insertOrDropItem(ItemStack stack) {
        assert level != null;
        var rest = InvUtils.injectToNearTile(level, getBlockPos(), stack);
        if (rest.isEmpty()) return; // Inserted all items to inventory.
        Containers.dropItemStack(level, getBlockPos().getX(), getBlockPos().getY(), getBlockPos().getZ(), rest);
    }

    @Override
    public void saveNbtData(CompoundTag nbt) {
        nbt.putBoolean("rs", rs);
        if (area != null)
            nbt.put("area", area.toNBT());
        var enchantments = new CompoundTag();
        this.enchantments.forEach(e ->
            enchantments.putInt(String.valueOf(e.enchantmentID()), e.level()));
        nbt.put("enchantments", enchantments);
        if (targetIterator != null)
            nbt.putLong("current", targetIterator.peek().asLong());
        nbt.put("inventory", container.createTag());
        nbt.put("denyList", denyList.stream().map(BlockStatePredicate::toTag).collect(Collectors.toCollection(ListTag::new)));
        nbt.put("allowList", allowList.stream().map(BlockStatePredicate::toTag).collect(Collectors.toCollection(ListTag::new)));
    }

    @Override
    public void load(CompoundTag nbt) {
        super.load(nbt);
        rs = nbt.getBoolean("rs");
        setArea(Area.fromNBT(nbt.getCompound("area")).orElse(null));
        var enchantments = nbt.getCompound("enchantments");
        setEnchantments(enchantments.getAllKeys().stream()
            .mapMulti(MapMulti.getEntry(ForgeRegistries.ENCHANTMENTS, enchantments::getInt))
            .map(EnchantmentLevel::new)
            .sorted(EnchantmentLevel.QUARRY_ENCHANTMENT_COMPARATOR)
            .toList());
        if (nbt.contains("current") && targetIterator != null)
            targetIterator.setCurrent(BlockPos.of(nbt.getLong("current")));
        container.fromTag(nbt.getList("inventory", Tag.TAG_COMPOUND));
        denyList = Stream.concat(nbt.getList("denyList", Tag.TAG_COMPOUND).stream()
            .mapMulti(MapMulti.cast(CompoundTag.class)).map(BlockStatePredicate::fromTag), defaultBlackList().stream()).collect(Collectors.toSet());
        allowList = nbt.getList("allowList", Tag.TAG_COMPOUND).stream()
            .mapMulti(MapMulti.cast(CompoundTag.class)).map(BlockStatePredicate::fromTag).collect(Collectors.toSet());
    }

    @Override
    public List<? extends Component> getDebugLogs() {
        return Stream.of(
            "%sArea:%s %s".formatted(ChatFormatting.GREEN, ChatFormatting.RESET, area),
            "%sTarget:%s %s".formatted(ChatFormatting.GREEN, ChatFormatting.RESET, Optional.ofNullable(targetIterator).map(MiniTarget::peek).orElse(null)),
            energyString()
        ).map(Component::literal).toList();
    }

    public void setEnchantments(List<EnchantmentLevel> enchantments) {
        this.enchantments = enchantments;
    }

    @Override
    public List<EnchantmentLevel> getEnchantments() {
        return enchantments;
    }

    Container getInv() {
        return container;
    }

    @Override
    public Component getDisplayName() {
        return getBlockState().getBlock().getName();
    }

    @Override
    public AbstractContainerMenu createMenu(int id, Inventory inv, Player player) {
        return new MiniQuarryMenu(id, player, getBlockPos());
    }

    static int interval(int efficiency) {
        return switch (efficiency) {
            case 0 -> 40;
            case 1 -> 30;
            case 2 -> 20;
            case 3 -> 10;
            case 4 -> 5;
            case 5 -> 2;
            default -> {
                if (efficiency < 0) yield 100;
                else yield 1;
            }
        };
    }

    static Set<BlockStatePredicate> defaultBlackList() {
        return Set.of(BlockStatePredicate.air(), BlockStatePredicate.fluid());
    }

    static boolean canAddInList(boolean isAllowList, BlockStatePredicate newData) {
        if (isAllowList) {
            return newData != BlockStatePredicate.all() && newData != BlockStatePredicate.air();
        } else {
            return !defaultBlackList().contains(newData);
        }
    }
}
